### PR TITLE
AsyncLocal  part 2 - replace all ThreadStatics and remove NodeCode and node CE

### PR DESCRIPTION
**Pull request by majocha**
This supersedes or follows up on #16602

Introduction of Transparent Compiler revealed some instability w.r.t. `DiagnosticsThreadStatics`:
https://github.com/dotnet/fsharp/pull/16576#issuecomment-1910321553
#16576
#16589

Turns out we used `[<ThreadStatic>]` to keep a compilation-global `DiagnosticsLogger`, and a custom `NodeCode` CE to flow it across threads during execution.
`AsyncLocal` is a BCL class that provides exactly that functionality: flowing state along the async execution path.

This replaces the `ThreadStatic` diagnostics state and Cancellable state with `AsyncLocal`.

`NodeCode<'T>` is no longer needed, so we remove it.

However `NodeCode` had some methods with subtly different behavior than `Async`.
Those are reproduced as Async type  extensions in `BuildGraph.fs`.

* `NodeCode.Sequential` ->`Async.SequentialImmediate`. (Because normal `Async.Sequential` queues tasks in the thread pool).
* `NodeCode.RunImmediateWithoutCancellation` ->`Async.RunImmediateWithoutCancellation`.

Some NodeCode methods effectively created a compilation scope, restoring previous global DiagnosticsLogger on finish:

* `NodeCode.FromCancellable`
* `NodeCode.AwaitAsync`
* `NodeCode.AwaitTask`

This doesn't seem to be covered by the test suite at the moment. (All the tests pass with or without) but  the following extensions wrap the computation in `use _ =  new CompilationGlobalsScope(...` preserving and restoring globals:
*  `Async.FromCancellableWithScope` 
* `Async.CompilationScope`

### CODE REVIEW

File: tests/FSharp.Compiler.UnitTests/BuildGraphTests.fs [hunk 1]
```diff
@@ -112,95 +113,87 @@ module BuildGraphTests =
...
-            try
-                NodeCode.RunImmediate(graphNode.GetOrComputeValue(), ct = cts.Token)
-                |> ignore
-                failwith "Should have canceled"
-            with
-            | :? OperationCanceledException as ex ->
-                ex
+        Assert.ThrowsAnyAsync<OperationCanceledException>(fun () ->
+            Async.StartImmediateAsTask(graphNode.GetOrComputeValue(), cancellationToken = cts.Token)      
+        ) |> ignore
```
[the remark this answers is not in the document]
Comment (majocha) [reply tests/FSharp.Compiler.UnitTests/BuildGraphTests.fs:174]: We do check it throws OperationCanceledException.

**Comment by majocha** [comment]
Yes this is still stuck on the same problem in CI.
I looked for performance regressions, even tried to reqrite the locking in BuildGraph (cause maybe just maybe it exhausts allowed handles when executing those 10000 simultanous `GetOrComputeValue`).

It looks like it freezes on a certain test in Linux / MacOS and something wierd also happens on Windows.

File: tests/FSharp.Compiler.UnitTests/BuildGraphTests.fs (hunk 1 above)
[the remark this answers is not in the document]
Comment (auduchinok) [reply tests/FSharp.Compiler.UnitTests/BuildGraphTests.fs:174]: Oops, right, thanks!

[the remark this answers is not in the document]
Comment (majocha) [reply tests/FSharp.Compiler.UnitTests/BuildGraphTests.fs]: No, thanks for looking at the code! I improved it a little bit now.

File: src/Compiler/Facilities/BuildGraph.fs [hunk 2]
```diff
@@ -225,87 +36,86 @@ module GraphNode =
...
-        // fast path
-        if isCachedResultNodeNotNull () then
-            cachedResultNode
-        else
-            node {
-                Interlocked.Increment(&requestCount) |> ignore
-
-                try
-                    let! ct = NodeCode.CancellationToken
+    member this.GetOrComputeValue() =
```
Comment (0101) [inline src/Compiler/Facilities/BuildGraph.fs:75]: Perhaps you pruned this code a little too much? Why not start with keeping it as-is just without `node`?

**Comment by majocha** [comment]
I don't know what to make of the CI logs. It seems they stop each time at the same spot.
If it was just general slowness introduced by this PR, I would expect it to more randomly timeout.

From just the logs I can't say what's going wrong here.

**Comment by majocha** [comment]
Where can I find the test that outputs stuff like 
```
Import (ReferenceArchAndPlatformArchMismatch: anycpu/anycpu) -- passed
Import (ReferenceArchAndPlatformArchMismatch: anycpu/x86) -- passed
Import (ReferenceArchAndPlatformArchMismatch: anycpu/x64) -- passed
Import (ReferenceArchAndPlatformArchMismatch: anycpu/Itanium) -- passed
Import (ReferenceArchAndPlatformArchMismatch: x86/anycpu) -- passed
Import (ReferenceArchAndPlatformArchMismatch: x86/x86) -- passed
Import (ReferenceArchAndPlatformArchMismatch: x86/x64) -- passed
Import (ReferenceArchAndPlatformArchMismatch: x86/Itanium) -- passed
Import (ReferenceArchAndPlatformArchMismatch: x64/anycpu) -- passed
Import (ReferenceArchAndPlatformArchMismatch: x64/x86) -- passed
Import (ReferenceArchAndPlatformArchMismatch: x64/x64) -- passed
Import (ReferenceArchAndPlatformArchMismatch: x64/Itanium) -- passed
Import (ReferenceArchAndPlatformArchMismatch: Itanium/anycpu) -- passed
Import (ReferenceArchAndPlatformArchMismatch: Itanium/x86) -- passed
Import (ReferenceArchAndPlatformArchMismatch: Itanium/x64) -- passed
Import (ReferenceArchAndPlatformArchMismatch: Itanium/Itanium) -- passed
```

?

**Comment by vzarytovskii** [comment]
@majocha if you'd like to, we can merge the first one, and then postpone this one a bit, I can go and try continue replacing with cold task-based one.

**Comment by majocha** [comment]
@vzarytovskii I'm fine with it. I think as is, it still improves things with Transparent Compiler.

**Comment by majocha** [comment]
OK, so this is green.

It appears all of the problems with this stemmed from my oversight.

NodeCode had subtly different behavior for functions like `Sequential`. SImply replacing it with `Async.Sequential` introduced some weird test failures that made me scratch my head for a week.

@vzarytovskii I think we can after all take this PR in favor of the #16602

I'll iron things out later. The diff is not small but reasonably easy to review.

**Comment by vzarytovskii** [comment]
That looks great! I want to have more extensive testing for this though, since it affects all compilation paths now, and technically it can affect a bunch of compilation flows. I trust that we can cover majority of those by just building our proto compiler and running tests, but would love to hear more opinions on that.

Any thoughts on that @0101, @majocha ?

**Comment by majocha** [comment]
Yes, I've seen first hand how the most subtle changes here were blowing up or freezing fsharpqa tests. It was eventually all caught by the tests, even if sometimes with little or no indication what's wrong.

So at the moment I have a lot of trust in the test suite. With the exception of the newest stuff:

For example the sporadic NRE in VS when using Transparent Compiler: https://github.com/dotnet/fsharp/pull/16576#issuecomment-1910321553

I'm pretty convinced this PR fixes it, but we really should have a repro and a test for this.

**Comment by 0101** [comment]
I'm a big fan of getting this in. I'll try to think of some more tests. I'd like to come up with some fuzzing test that would reveal the NRE we were seeing. 

Also, should we try to run some benchmarks on this change to see if performance has changed? There's now less wrapping but also `AsyncLocal` could have some costs.

**Comment by majocha** [comment]
Yes for seriously benchmarking this. I've run only FCSSourceFiles from time to time and the final version seems to run the same as `main`.

**Comment by majocha** [comment]
I wonder if uncommenting these asserts would reveal some potential problem points, anything not caught by the test suite. This todo is there probably since 2017:
https://github.com/dotnet/fsharp/blob/589d350ac2831cb6d52049eeb4118a1af35bf9e1/src/Compiler/Facilities/DiagnosticsLogger.fs#L345-L350

I guess the original idea here was that `AssertFalseDiagnosticsLogger` would signal we are pushing diagnostics but have no initialized any DiagnosticLogger.

`assert` works only in debug builds?

**Comment by 0101** [comment]
Yeah that might be a good idea. We can also just put `failwith` there.

**Comment by majocha** [comment]
Yeah, the only problem is Transparent Compiler triggers that exception also in VS, not just in the test.

**Comment by 0101** [comment]
Good catch. This needs to be fixed probably in both places:

- Don't try to replay diagnostics to a Null Logger (though it would be nice to send this to telemetry somehow)
- Don't run (request) a computation that produces diagnostics without having set up a logger - wonder where we do that

**Comment by majocha** [comment]
> * Don't run (request) a computation that produces diagnostics without having set up a logger - wonder where we do that

There may be a difference in behavior between the old builder and Transparent Compiler. Possibly there are places where the old one sets a global logger and doesn't bother to restore previous one later.

**Comment by majocha** [comment]
I think I may have also overlooked another thing here. `NodeCode.FromCancellable` and `NodeCode.AwaitTask` effectively wrapped the computation in a compilation diagnostics scope (restoring the global logger and build phase on finish). `Async.AwaitNodeCode` did the same thing.
There may subtle changes in behavior here that pass the test suite anyway.

**Comment by majocha** [comment]
> * Don't try to replay diagnostics to a Null Logger (though it would be nice to send this to telemetry somehow)

I included a provisional fix for this in the commit above.

**Comment by 0101** [comment]
> There may be a difference in behavior between the old builder and Transparent Compiler. Possibly there are places where the old one sets a global logger and doesn't bother to restore previous one later.

Yeah it's quite different. But there are not that many places where we call `AsyncMemoize.Get`. So we can still check if some of them can produce diagnostics and there's no logger above.

**Comment by majocha** [comment]
OK, with the above changes I have one failure in fsharpqa locally.

**Comment by majocha** [comment]
Looks like it's still good. I installed the above fix in VS and Transparent Compiler works.

**Comment by majocha** [comment]
Another diagnostics push to null, find usages with Transparent Compiler in VS:
>   at FSharp.Compiler.DiagnosticsLogger.UninitializedDiagnosticsLogger@347.DiagnosticSink(PhasedDiagnostic diagnostic, FSharpDiagnosticSeverity severity) in E:\repos\fsharp\src\Compiler\Facilities\DiagnosticsLogger.fs:line 349
   at FSharp.Compiler.CompilerOptions.reportDeprecatedOption@367-1.Invoke(FSharpOption`1 errOpt) in E:\repos\fsharp\src\Compiler\Driver\CompilerOptions.fs:line 369
   at FSharp.Compiler.CompilerOptions.attempt@373.Invoke(FSharpList`1 l) in E:\repos\fsharp\src\Compiler\Driver\CompilerOptions.fs:line 379
   at FSharp.Compiler.CompilerOptions.processArg@333.Invoke(FSharpList`1 args) in E:\repos\fsharp\src\Compiler\Driver\CompilerOptions.fs:line 503
   at FSharp.Compiler.CompilerOptions.ParseCompilerOptions(FSharpFunc`2 collectOtherArgument, FSharpList`1 blocks, FSharpList`1 args) in E:\repos\fsharp\src\Compiler\Driver\CompilerOptions.fs:line 506
   at FSharp.Compiler.CompilerOptions.ApplyCommandLineArgs(TcConfigBuilder tcConfigB, FSharpList`1 sourceFiles, FSharpList`1 argv) in E:\repos\fsharp\src\Compiler\Driver\CompilerOptions.fs:line 2317

**Comment by majocha** [comment]
> For now we can probably just add a note about it and rely on your fix with discard error logger.

Hmm. This one doesn't seem to go through the cache.

**Comment by 0101** [comment]
Oh yeah, here where it says we might need to deal with exceptions? 😅 I guess we do...

https://github.com/dotnet/fsharp/blob/3ac064e3f838f66cdf8c2929aac129b981c68a23/src/Compiler/Service/TransparentCompiler.fs#L1801-L1802

**Comment by majocha** [comment]
We can do for now:
```fsharp
            // TODO: might need to deal with exceptions here:
            let tcConfigB, sourceFileNames, _ =
                DiagnosticsLogger.suppressErrorReporting <| fun () ->
                    ComputeTcConfigBuilder projectSnapshot
```

**Comment by majocha** [comment]
Probably these could be just functions in `DiagnosticsLogger.fs`:
*  `Async.FromCancellableWithScope` 
* `Async.CompilationScope`

As they have nothing to do with Async in general.

**Comment by majocha** [comment]
A quick `TransparentCompilerBenchmark` shows the performance is mostly unchanged.

```
asynclocal
| Method          | UseTransparentCompiler | ProjectType     | Mean       | Error     | StdDev    | Gen0       | Completed Work Items | Lock Contentions | Gen1      | Allocated |
|---------------- |----------------------- |---------------- |-----------:|----------:|----------:|-----------:|---------------------:|-----------------:|----------:|----------:|
| ExampleWorkflow | False                  | DependencyChain | 3,045.4 ms |  66.11 ms |  10.23 ms | 19000.0000 |             157.0000 |                - | 2000.0000 |   5.62 GB |
| ExampleWorkflow | False                  | DependentGroups | 3,050.4 ms |  43.11 ms |   6.67 ms | 19000.0000 |             150.0000 |                - | 2000.0000 |   5.58 GB |
| ExampleWorkflow | False                  | ParallelGroups  | 3,286.8 ms |  55.72 ms |   8.62 ms | 20000.0000 |             151.0000 |                - | 3000.0000 |   5.79 GB |
| ExampleWorkflow | True                   | DependencyChain | 2,287.5 ms | 190.75 ms |  10.46 ms |  2000.0000 |             862.0000 |                - | 1000.0000 |   2.84 GB |
| ExampleWorkflow | True                   | DependentGroups |   977.4 ms | 704.55 ms | 109.03 ms |  2000.0000 |             824.0000 |                - | 1000.0000 |   2.75 GB |
| ExampleWorkflow | True                   | ParallelGroups  |   909.6 ms | 660.61 ms | 102.23 ms |  2000.0000 |             859.0000 |                - | 1000.0000 |   2.98 GB |

main
| Method          | UseTransparentCompiler | ProjectType     | Mean       | Error    | StdDev    | Gen0       | Completed Work Items | Lock Contentions | Gen1      | Allocated |
|---------------- |----------------------- |---------------- |-----------:|---------:|----------:|-----------:|---------------------:|-----------------:|----------:|----------:|
| ExampleWorkflow | False                  | DependencyChain | 3,496.0 ms | 851.2 ms | 131.73 ms | 19000.0000 |             156.0000 |                - | 2000.0000 |   5.61 GB |
| ExampleWorkflow | False                  | DependentGroups | 3,330.6 ms | 196.5 ms |  30.40 ms | 19000.0000 |             151.0000 |                - | 2000.0000 |   5.58 GB |
| ExampleWorkflow | False                  | ParallelGroups  | 3,379.3 ms | 410.1 ms |  63.46 ms | 20000.0000 |             151.0000 |                - | 4000.0000 |   5.78 GB |
| ExampleWorkflow | True                   | DependencyChain | 2,328.7 ms | 493.9 ms |  27.07 ms |  2000.0000 |             839.0000 |                - | 1000.0000 |   2.84 GB |
| ExampleWorkflow | True                   | DependentGroups |   934.5 ms | 583.5 ms |  90.30 ms |  2000.0000 |             868.0000 |                - | 1000.0000 |   2.74 GB |
| ExampleWorkflow | True                   | ParallelGroups  |   951.4 ms | 799.7 ms | 123.76 ms |  2000.0000 |             901.0000 |                - | 1000.0000 |   2.98 GB |
```

**Comment by 0101** [comment]
We should probably figure out some more testing scenarios before we merge this.

**Comment by majocha** [comment]
Yep, also there's a fail in CI just now. Either a bad merge or indicative of something else.

**Comment by majocha** [comment]
I have a wacky idea. We could try to record changes of global (threadstatic) diagnostic logger from some compiler runs to get some kind of baseline. But that would not be deterministic. Not in a linear way, we would have to record all the forks in execution anytime there is some parallelization. That would give us some kind of tree-like baseline to compare against.

Another, simpler idea is to start clean from `main`,  put the AsyncLocal state alongside the current ThreadStatic, kind of shadowing it, and just throw or assert whenever they diverge. this one is easier. I kind of have an idea how to do it.

**Comment by 0101** [comment]
Good idea. And you pretty much have it ready. Please try it and let us know. I think we can just keep that in a separate branch.

**Comment by majocha** [comment]
Apart from `Cancellable`, done in #16779

**Pull request closed without merging**